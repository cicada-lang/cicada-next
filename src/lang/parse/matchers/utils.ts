import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
