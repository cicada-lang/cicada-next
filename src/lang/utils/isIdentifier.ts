const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
