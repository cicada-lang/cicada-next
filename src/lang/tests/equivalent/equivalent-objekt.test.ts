import { expect, test } from "vitest"
import { expectCodeToFail, runCode } from "../utils.js"

test("equivalent Objekt", async () => {
  const output = await runCode(`

function id(T: Type, x: T): T {
  return x
}

compute equivalent class { A: Type, x: A } {
    { A: Trivial, x: sole }
  = { A: Trivial, x: sole }
  = { A: Trivial, x: id(Trivial, sole) }
}

`)

  expect(output).toMatchSnapshot()
})

test("equivalent Objekt -- fail", async () => {
  await expectCodeToFail(`

compute equivalent class { A: Type, x: A, y: A } {
    { A: String, x: "abc", y: "xyz" }
  = { A: String, y: "xyz", x: "abcd" }
}

`)

  await expectCodeToFail(`

function id(T: Type, x: T): T {
  return x
}

compute equivalent class { A: Type, x: A, y: A } {
    { A: String, x: id(String, "abcd"), y: id(String, "xyz") }
  = { A: String, y: id(String, "xyz"), x: id(String, "abc") }
}

`)
})
