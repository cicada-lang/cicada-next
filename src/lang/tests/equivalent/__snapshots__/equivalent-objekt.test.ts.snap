// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`equivalent Objekt > 
    "refl(
      implicit class {
        A: Type
        x [rename: x2]: A
      },
      implicit {
        A: Trivial,
        x: sole
      },
    ): Equal(
      class {
        A: Type
        x [rename: x2]: A
      },
      {
        A: Trivial,
        x: sole
      },
      {
        A: Trivial,
        x: sole
      },
    )"
   1`] = `
"refl(
  implicit class {
    A: Type
    x [rename: x2]: A
  },
  implicit {
    A: Trivial,
    x: sole
  },
): Equal(
  class {
    A: Type
    x [rename: x2]: A
  },
  {
    A: Trivial,
    x: sole
  },
  {
    A: Trivial,
    x: sole
  },
)"
`;

exports[`equivalent Objekt: 
    "refl(
      implicit class {
        A: Type
        x [rename: x2]: A
      },
      implicit {
        A: Trivial,
        x: sole
      },
    ): Equal(
      class {
        A: Type
        x [rename: x2]: A
      },
      {
        A: Trivial,
        x: sole
      },
      {
        A: Trivial,
        x: sole
      },
    )"
   1`] = `
"refl(
  implicit class {
    A: Type
    x [rename: x2]: A
  },
  implicit {
    A: Trivial,
    x: sole
  },
): Equal(
  class {
    A: Type
    x [rename: x2]: A
  },
  {
    A: Trivial,
    x: sole
  },
  {
    A: Trivial,
    x: sole
  },
)"
`;

exports[`equivalent Objekt 1`] = `
"refl(
  implicit class {
    A: Type
    x [rename: x2]: A
  },
  implicit {
    A: Trivial,
    x: sole
  },
): Equal(
  class {
    A: Type
    x [rename: x2]: A
  },
  {
    A: Trivial,
    x: sole
  },
  {
    A: Trivial,
    x: sole
  },
)"
`;
